src/test/styles/evaluators-spec.ts

```diff
@@ -0,0 +1,294 @@
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CalcEvaluator, EnvEvaluator, PercentageEvaluator, OperatorEvaluator, StyleEvaluator, SphericalIntrinsics} from '../../styles/evaluators.js';
+import {expressionNode, functionNode, identNode, operatorNode, spy} from '../helpers.js';
+import { numberNode } from '../../styles/parsers.js';
+
+const expect = chai.expect;
+
+suite('evaluators', () => {
+  suite('EnvEvaluator', () => {
+    test('is never constant', () => {
+      const evaluator = new EnvEvaluator(functionNode('env', []));
+      expect(evaluator.isConstant).to.be.false;
+    });
+
+    test('with no arguments, evaluates to zero', () => {
+      const evaluator = new EnvEvaluator(functionNode('env', []));
+      expect(evaluator.evaluate()).to.be.eql(numberNode(0, null));
+    });
+
+    suite('window-scroll-y', () => {
+      test('evaluates to current top-level scroll position', () => {
+        const evaluator = new EnvEvaluator(functionNode(
+            'env', [expressionNode([identNode('window-scroll-y')])]));
+        expect(evaluator.evaluate()).to.be.eql(numberNode(0, null));
+        const scrollPosition = 10000;
+        const scrollMax = 20000;
+
+        const restorePageYOffset =
+            spy(window, 'pageYOffset', {value: scrollPosition});
+        const restoreBodyScrollHeight =
+            spy(document.documentElement, 'clientHeight', {value: scrollMax});
+
+        expect(evaluator.evaluate())
+            .to.be.eql(numberNode(
+                scrollPosition / (scrollMax - window.innerHeight), null));
+
+        restorePageYOffset();
+        restoreBodyScrollHeight();
+      });
+    });
+  });
+
+  suite('PercentageEvaluator', () => {
+    test('multiplies the percentage by the basis', () => {
+      const evaluator = new PercentageEvaluator(numberNode(200, '%'), numberNode(1, 'm'));
+      expect(evaluator.evaluate()).to.be.eql(numberNode(2, 'm'));
+    });
+  });
+
+  suite('CalcEvaluator', () => {
+    test('is constant if its operands are all numbers', () => {
+      const evaluator = new CalcEvaluator(functionNode(
+          'calc',
+          [expressionNode(
+              [numberNode(1, null), operatorNode('+'), numberNode(1, null)])]));
+
+      expect(evaluator.isConstant).to.be.true;
+    });
+
+    test('is constant if nested functions are constant', () => {
+      const evaluator = new CalcEvaluator(functionNode(
+          'calc', [expressionNode([
+            numberNode(1, null),
+            operatorNode('+'),
+            functionNode('calc', [expressionNode([numberNode(1, null)])])
+          ])]));
+
+      expect(evaluator.isConstant).to.be.true;
+    });
+
+    test('is not constant if any nested function is not constant', () => {
```

Comment:
What if you have a percentage of window-scroll-y? I saw a comment earlier that makes me think this would return const. Also, does percentage of window-scroll-y actually make any sense, since it seems redundant?

Reply:
Mathematically it makes sense. Say I have the following expression: `calc(50% + 50% * env(window-scroll-y))`.

For a basis of `1m` this would work out to `0.5m` at the minimum scroll position and `1m` at the maximum scroll position.